Rich-Text-Editor-(RTE)-Overview.md

```diff
@@ -1,117 +1,137 @@
+2. Add a call to `CKEditor.plugins.addExternal` to the component's `ngAfterViewInit()` method.
+3. Add this plugin to `extraPlugins` in `ckConfig`.
+
+#### Adding rich text components to CKEditor
+
+Rich text components are added to CKEditor in [`ck-editor-4-widgets.initializer.ts`](https://github.com/oppia/oppia/blob/develop/core/templates/components/ck-editor-helpers/ck-editor-4-widgets.initializer.ts). The components are dynamically added to CKEditor as [widgets](https://docs.ckeditor.com/ckeditor4/latest/guide/widget_sdk_intro.html). We use the `getRichTextComponents()` function to obtain the information of each rich text component and construct there respective widgets. We also have a function `isInlineComponent()` to check whether a rich text component is inline or block component.
+
+`componentTemplate` defines a template to wrap the rich text components. Inline components are wrapped in a span, and block components are wrapped in a div.
+
+The plugins are added to CKEditor in the line `CKEDITOR.plugins.add(ckName, {`. The `init` function passed to `add()` is executed when the plugin is initialised and it adds a widget for each component using the line `editor.widgets.add(ckName, {`. `editor.widgits.add` is passed an object with the following attributes:
+
+* `button`: The button label for the widget.
+* `inline`: Whether the widget is an inline component or block component. The link and the math component are inline and the rest are declared block components.
+* `template`: The wrapper template for the widget.
+* `edit`: A function that will be executed when a widget is being edited. In this function the default action is canceled since we have used our own edit modal. `RteHelperService._openCustomizationModal()`is a helper function for opening the modal which is used to insert new components or editing existing ones. It uses the customizationArgSpecs (obtained from the component definition) to know what the editable fields are for each component, which allows it to render the modal properly for any component.
+* `downcast()`: A function that downcasts the widget instance by clearing the Angular rendering content and returning the rich text component without any wrapper. Downcasting a widget means removing any property which is not needed in the output's source.
+* `upcast()`: A function that upcasts an element to this widget. It returns whether an element is an instance of the widget. The element will be upcasted if it is an instance of the widget. Upcasting an element means changing that element into a widget.
+* `data()`: A function that will be executed every time the widget data changes. It will set the attributes of rich text components according to the change in data values.
+* `init()`: A function that will be executed when initialising a widget (after a widget instance is created, but before it is ready). It is executed before the first time when data function is exceuted. It reads and saves values from the component attributes.
```

Comment:
after --> i.e. after

This description may need some proofreading.

Reply:
done